specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,501 @@
+It depends on the hash/address/node of the items and a 'chunking factor' for tuning the shape of the tree.
+This chunking factor determines the average size of nodes/chunks that a node on a higher level contains.
+This average size in turn controls the shape of the tree (broad/narrow).
+The shape in turn defines the performance of operations on the tree with the following tradeoffs:
+  - larger blocks with more cache invalidations
+  - or smaller blocks with less cache invalidations and more lookups
+
+### Note on Multiple and Single Values and Sets
+
+The described tree can represent a data structure with multiple, single or no values per key. However, given that IPLD Maps (which a Prolly Tree loosely maps to) only allow one value for a key, implementations should merge duplicate keys into a single value. This is also important to have consistent ordering of key value pairs.
+
+### Rebalance
+
+In order to modify the tree (inserting, updating, or removing one or multiple values) a part of the tree needs to be rebalanced. This section describes an internal procedure that is not exposed to the caller.
+The tree is (partially) rebalance bottom-up.
+
+Starting at the inserted/modified leaf or node referencing a now deleted/removed leaf, successively walk up the tree. If a removed item was the only item in its node: Remove the node from its parent and Check if the hash/address/CID of a new leaf/node splits the parent chunk (split the parent node if yes).
+Then, check if a removed leaf/node was a 'splitting node' and the nodes need to be merged (only the last node in a chunk can be a splitting node). If splitting criterion holds for the last item of the node and there is the current node is succeeded by another node on the same level: merge the current node with the succeeding node. Continue walking up the path. If at root and it is being split: Create a new root that links to the freshly split nodes. Return the new root CID.
+
+Pay attention to the fact that the boundary algorithm is not related to the node CID, i.e the boundary is generated *before* the node CID is generated.
+
+### Put/Remove/Update
+
+First search if the tree already contains the value.
+If the key is not contained, create a new leaf for the value, insert the key-value (key-leaf) pair in the correct node and rebalance the tree.
+If the key is already stored in the tree and the value is the same, the tree is left unchanged.
+If the key exists and the values differ: Either leave the tree or insert the new value and update the tree.
+If removing the key, remove the `key` and the `value` from the list. After making any modifications to the tree, run the (Rebalance)[#Rebalance] algorithm.
+
+## Structure
+
+```
+type ProllyTree struct {
+  config &ProllyTreeConfig
+  root &TreeNode
+} representation tuple
+
+type TreeNode struct {
+  # Is leaf when level is 0
+  level Int
```

Comment:
I really don't think the level should go into the data.

It can't be trusted from the point of the view of the reader, and for the writer who is using this info to determine chunker settings they'll always know the depth because you've gotta write it from the branches up.

You do still need a way to differentiate branch from leaf nodes though, either as a type difference or a boolean in the data.